apps/src/lib/node/ledger/shell/queries.rs

```diff
@@ -604,41 +596,43 @@ where
     }
 
     #[cfg(feature = "abcipp")]
-    fn can_send_validator_set_update(&self, can_send: SendValsetUpd) -> bool {
-        let (check_prev_heights, height) = match can_send {
-            SendValsetUpd::Now => (false, self.get_current_decision_height()),
-            SendValsetUpd::AtPrevHeight => (false, self.last_height),
-            SendValsetUpd::AtFixedHeight(h) => (true, h),
-        };
+    fn can_send_validator_set_update(&self, _can_send: SendValsetUpd) -> bool {
+        // TODO: implement this method for ABCI++; should only be able to send
+        // a validator set update at the second block of an epoch
+        true
+    }
 
-        // handle genesis block corner case
-        if height == BlockHeight(1) {
+    #[cfg(not(feature = "abcipp"))]
+    fn can_send_validator_set_update(&self, can_send: SendValsetUpd) -> bool {
+        // when checking vote extensions in Prepare
+        // and ProcessProposal, we simply return true
+        if matches!(can_send, SendValsetUpd::AtPrevHeight) {
             return true;
         }
```

Comment:
We can get rid of the `SendValsetUpd` now, we no longer need this context from the caller? `can_send_validator_set_update` can reason by itself using only the current decision height.

Reply:
we call this from two places:

* when we create a new vote extension at the ABCI commit phase, where we pass `SendValsetUpd::Now` as arg
* when we verify vote extensions e.g. at prepare and process proposal, where we pass `SendValsetUpd::AtPrevHeight` as arg

we still need to distinguish between these two cases; a `BlockHeight` parameter wouldn't be enough, since we need `SendValsetUpd::AtPrevHeight` to always return true.

Reply:
you could remove it altogether if we refactored prepare and process proposal to always accept validator set update protocol txs without a predicate check, but then it'd be kind of annoying to write this code again for ABCI++ if/when that happens...

Reply:
> you could remove it altogether if we refactored prepare and process proposal to always accept validator set update protocol txs without a predicate check, but then it'd be kind of annoying to write this code again for ABCI++ if/when that happens...

I would actually do this but since we're going to maintain our current `abcipp` code I guess it's fine